### ISSUE TITLE
`_encode_prompt` function

### ISSUE THREAD

**Issue by patrickvonplaten** [huggingface/diffusers#4140]
At the moment, we have a bit of an inconsistent design with the `_encode_prompt` / `encode_prompt` functions.

We essentially have three different designs:

- 1.) The original stable diffusion 1.5 design, which means we have a private `_encode_prompt` function that does **not** use by default use `torch.no_grad()` and returns a single tensor
https://github.com/huggingface/diffusers/blob/6427aa995e0c03c1c1a635cd6af8e365f47541a8/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L302

- 2.) For IF we made the method public and wrapeed it into a `torch.no_grad()` mainly because one more or less has to run this function independently because of the size of the text encoder and the multi-stage inference process
https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/deepfloyd_if/pipeline_if.py#L246

- 3.) For SD-XL for now we also made the method public, but did **not** add a `torch.no_grad()` decorator. The reason being here that there might be use cases where the user actually wants to compute the gradients when calling `encode_prompt` - e.g. when training LoRA with text encoder one could just call encode_prompt
https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L212

What are our options to improve the API / unify the API  here? cc @yiyixuxu @sayakpaul @williamberman @pcuenca

**Comment by yiyixuxu** [comment]
I think it makes sense to make `encode_prompt` a public method and wrapped inside `@torch_no_grad()` across all our pipelines - it might make it less convenient for training but I think it's ok to be a little bit more verbose in our training scripts.

**Comment by eliphatfs** [comment]
Though this may be a bit off-topic, I actually want the different parts in pipeline to be callable (and without `no_grad`) for pipelines so in training scripts we can reuse them and we don't have that many duplicate logic and worries that the training is actually off against the inference.

**Comment by eliphatfs** [comment]
Though this may be a bit off-topic, I actually want the different parts in pipeline to be callable (and without `no_grad`) for pipelines so in training scripts we can reuse them and we don't have that many duplicate logic and worries that the training is actually off against the inference.

**Comment by sayakpaul** [comment]
Yeah I can understand that concern. This is why I think the utils what we are brainstorming here could make sense.

**Comment by patrickvonplaten** [comment]
So, first it seems like we should make a public `encode` method that does **not** have a `torch.no_grad()` dependency so that it can be used for training (so essentially what we're doing right now for SD-XL) => this makes sense to me.

Regarding a utils encoder class, I have the following concerns:
- encoding can be **very** pipeline-specific. E.g. IF and Stable Diffusion XL have their very own way of encoding. It would not make sense to build a class for them IMO. Also even classes like Stable Diffusion Instruct Pix2Pix don't have the same encoding class as Stable Diffusion
- Creating a "encoding" class would entail having to save the text encoder and tokenizer as attributes of this class which means that we would have both `pipe.text_encoder` and `pipe.encoder.text_encoder` assuming that encoder is such a utils class => this is not great as the user now has a more complex way of accessing tokenizer and encoder

I do admit that there is quite some boiler plate encoding code between SD pipeline classes though. 

Do you have any design ideas how we could:
- a) reduce boiler plate code without sacrificing simplicity for the reader
- b) have a design that does not wrap text_encoder and tokenizer into a new class

maybe also cc @williamberman

**Comment by sayakpaul** [comment]
Hmm, fair enough. 

I think this relates to the philosophy of Transformers. Duplication of code to promote readability. The subtle differences in some of the text encoding schemes also make it hard to approach this refactor.

**Comment by williamberman** [comment]
For the gradient question, this is the code I'd recommend. If we want it to be used in the training scripts, I'd also recommend moving encode_prompt_with_grad to a class function so we don't have to instantiate a pipeline. I think the exposed encode_prompt function should always be wrapped in no_grad because the pipelines are primarily used for inference.

```
class Pipeline
    @no_grad
    def encode_prompt():
        return self.encode_prompt_with_grad(...)

    @classmethod
    def encode_prompt_with_grad():
        ....
```

For boiler plate reduction, I don't know off the top of my head, would have to do some thought :)

**Comment by patrickvonplaten** [comment]
I don't think `encode_prompt_with_grad()` should be a class method - I don't want to pass the encoder etc... to it.
Encode prompt is too tied to the components of a pipeline.
 
I'd prefer to just decided to **not** add `torch.no_grad()` & remove it from IF. This is also very much in line with PyTorch and keeping this library flexible so that encode_prompt can be used for training.

**Comment by patrickvonplaten** [comment]
So I'd be in favor of aligning everything to 3.) - happy to open a PR for this

**Comment by sayakpaul** [comment]
To summarize, it's hard to completely disentangle `encode_prompt()` from the pipelines. So, I think it's best to give each pipeline its `encode_prompt()` with clear `# Copied from ...` statements to help with maintenance. 

Removal of `torch.no_grad()` should be useful. This way we also stay explicit similar to how we deal with the placement of modules on a requested device. 

Am I on the same page as you?

**Comment by patrickvonplaten** [comment]
Yes, I'd be in favor of this!

**Comment by williamberman** [comment]
How many docs do we have already released that recommend using `encode_prompt` to re-use encode prompts? My understanding was it's considered a public api to be used in inference. I'd worry about removing no grad for it now.

I don't think I necessarily agree that encode_prompt is very dependent on the pipeline class. It doesn't mutate any state on the pipeline class, and it only requires reading two instance variables (the text encoder and the tokenizer). 

If the recommended api to use the encoding from the training script is to instantiate a pipeline, we then have to manually pass None to all the other constructor arguments or call from_pretrained. We've already had success in simplifying the lora loader script where we moved instance methods to class methods where we are unecessarily calling from_pretrained and downloading weights just to call a function.

i.e.
```python
tokenizer = ...
text_encoder = ...

pipe = Pipeline(tokenizer=tokenizer, text_encoder=text_encoder, unet=None, safety_checker=None, <... other arguments might be none here>)
pipe.encode_prompt("foo bar")
```

compare this to 

```py
tokenizer = ...
text_encoder = ...

pipe = Pipeline.encode_prompt("foo bar", tokenizer, text_encoder)
```

I'm pro moving as many things as possible away from instance methods if they don't mutate instance state.

**Comment by eliphatfs** [comment]
What are the actual cases where you have a tokenizer and a text encoder but not other parts of the pipeline? I prefer constructing a pipeline and simply use it for everything including training.
The argument on `no_grad` is valid. Perhaps we have to add other APIs for use without `no_grad`.

**Comment by patrickvonplaten** [comment]
> How many docs do we have already released that recommend using encode_prompt to re-use encode prompts? My understanding was it's considered a public api to be used in inference. I'd worry about removing no grad for it now.

Just to clarify it's only IF's encode_prompt that has a `torch.no_grad()`. We could leave it there indeed for backward compatibility. For all other encode_prompts we should **not** add it IMO so that we render this method useful for training as well. This is in my opinion also more in sync with PyTorch API.

> I don't think I necessarily agree that encode_prompt is very dependent on the pipeline class. It doesn't mutate any state on the pipeline class, and it only requires reading two instance variables (the text encoder and the tokenizer).

Hmm all the following have different pipelines:
- IF (weird pre-processing)
- Kandinsky (text_emb -> image_emb)
- SD (no attention mask)
- Alt Diffusion (with attention mask)
- SD XL (two tokenizers) 
- ...

> I'm pro moving as many things as possible away from instance methods if they don't mutate instance state.
This is not a very PyTorch-y design, that's much more a Flax/JAX design where we always try to stay function and make a big different between mutating state and not. PyTorch is not designed around state mutation - we also don't pass weights to a PyTorch forward function. `encode_prompt` of a pipeline should only ever be used with the text encoder of the pipeline and there is no need to allow other text encoders to be passed. The end-user has to then pretty much know what to pass to encode prompt (which would be two text encoders & tokenizers for SDXL, which is too much).

**Comment by williamberman** [comment]
> It's only IF's encode_prompt that has a `torch.no_grad()`. We could leave it there indeed for backward compatibility. For all other encode_prompts we should **not** add it IMO so that we render this method useful for training as well. This is in my opinion also more in sync with PyTorch API.

Ah if only it's only IF, then I'm ok with changing it. I was under the impression it was more standardly used :) 


> Hmm all the following have different pipelines:
> 
> * IF (weird pre-processing)
> * Kandinsky (text_emb -> image_emb)
> * SD (no attention mask)
> * Alt Diffusion (with attention mask)
> * SD XL (two tokenizers)

Well in all these cases the _logic_ of the function is different but still none are dependent on instance state which was the main point I was getting at. I can't really think of text encoding logic that would require holding instance state.


> This is not a very PyTorch-y design, that's much more a Flax/JAX design where we always try to stay functional and where mutating / not mutating state is a fundamental design of the language.

I'd consider this to actually be relatively close to how most (all?) pytorch classes actually wrap functional equivalents that support more customization (and are public apis). I.e. when we want the easier to use api and don't care about maintaining our own state, we use nn.Conv2d/the instance method -- and then when we want more control over the inputs, we call F.conv2d/the class method. I don't necessarily see it as all or nothing :) 


> If we would make it a class method, the end-user has to then pretty much know what to pass to encode prompt (which would be two text encoders & tokenizers for SDXL), which makes things unnecessarily complex. If encode_prompt shall be used for training it is indeed easier to have it be a class method. But I don't think this warrants it to make it a class method since 99% of use cases are inference. 

I'm not recommending making it _only_ a class method :) My original code snippet contains the logic in a class method which delegates to a class method which contains most of the logic. Imo this is the best of both worlds since for the standard use case, inference which always occurs on a pipeline instance, the user can just call the instance method without passing any parameters in. But for training, we can call the method without instantiating a pipeline.

> Also there is no use case where the unet would not be trained as far as I can see.

Yes but many pipeline have more than just a text encoder/tokenizer/unet. If we have a safety checker, prior, etc... we will always have to remember how to instantiate the class in order to call a method that doesn't rely on those instance variables.


I'm happy with whatever gets chosen here as I think any decreased ergonomics in writing training scripts are able to be worked around and the inference api should be prioritized :) 

I just think in general we can both maintain the simplicity of the stateful pipeline interface and the flexibility to use components in the training script through this pattern of instance method which delegates to static method similar to how we did for lora loading.

**Comment by patrickvonplaten** [comment]
Thanks for the discussion here! If ok for everybody I'd for now go with:

1. Leave IF's encode prompt as is
2. Add a new `encode_prompt` to all SD models **without** a `torch.no_grad()` and deprecate `_encode_prompt`
3. If desired, I'm fine with additionally adding a class method `encode_prompt_func` or something.

@sayakpaul would you maybe want to give this a try?
